### DISCUSSION TITLE
Enemies Lists Overhaul

### DISCUSSION THREAD

**Discussion by OcelotWalrus** [Dungeons-of-Kathallion/Bane-Of-Wargs#86] [Ideas]
Hey @lumbar527, I've gotten an idea recently, that'd drastically enhance the actual battle/fights in the currents versions of Bane Of Wargs. You see, in the currents versions, when enemies spawn, only a single type of enemy actually spawn (3 wargs), but it'd really be more interesting for the gameplay to make so that enemies could spawn in raids, where in each raid you can define the percentage of chance an enemy type has to spawn, and also a minimum and maximum number of enemies spawns, for every difficulty mode (like that, in easy mode, there's lower enemies rate and in heard mode there's higher enemies rate).

It'd look like this in my mind:

```python
[`lists.yaml` file]
generic:  # the generic list, with the new future data
  Goblin Raid:
    name: <name(str)>  # this would be the name displayed when it spawns ?
    chance: <percentage(float)>
    enemies rate:
      easy: {"min": 1, "max": 2}  # more compact dictionary so it takes lesser space
      normal: {"min": 1, "max": 4}
      hard: {"min": 2, "max": 5}
    enemies spawns:
      <enemy type>: <spawning chance(float)>
      <enemy type>: <spawning chance(float)>
      [...]
```

Instead of this in the actual game engine:

```python
[`lists.yaml` file]
generic:
- Goblin
- Goblin
- Ork
- Ork
- Ork
- Ork
- Ork Archer
- Ork Archer
- Ork Archer
- Warg
- Cavern Troll
- Slime
- Slime
```

And this's how it'd basically work, to create the random raid:

```python
# We assume that the player's playing normal difficulty
enemy_list = lists["generic"]
enemies_number = random.randint(
    enemy_list["enemies rate"]["normal"]["min"],
    enemy_list["enemies rate"]["normal"]["max"]
)
enemy_entries_length = len(list(enemy_list["enemies spawns"]))

enemies = []
count = 0
while count < enemies_number:
    for entry in enemy_list["enemies spawns"]:
        if count > enemies_number:  # I'm not sure if the loop will stop in the for loop or wait the for loop
            return
        enemy_rate = enemy_list["enemies spawns"][entry]
        if enemy_rate > random.uniform(0, 1):
            enemies += [entry]
        count += 1
````

---

I'd like to hear your comments and ideas about this new feature of the game engine!

**Comment by lumbar527** [comment]
Sounds interesting!
What if, instead of
```
      <enemy 1>:  # any name would work for the game engine
        type: Goblin
        spawning chance: .4  # 40% of spawning chance
```
the format was
```
      Goblin: .4
```
?

**Comment by OcelotWalrus** [comment]
> Sounds interesting! What if, instead of
> 
> ```
>       <enemy 1>:  # any name would work for the game engine
>         type: Goblin
>         spawning chance: .4  # 40% of spawning chance
> ```
> 
> the format was
> 
> ```
>       Goblin: .4
> ```
> 
> ?

Seems like you're smarter than me!

**Comment by OcelotWalrus** [comment]
@lumbar527, I'd also like to get your ideas on some new enemies, to enhance the current [enemy database](https://github.com/Dungeons-of-Kathallion/Bane-Of-Wargs/blob/master/data/enemies.yaml).

**Reply by lumbar527** [reply]
Maybe undead? Evil people? _Korath?_

**Reply by OcelotWalrus** [reply]
Undeads are a good idea! There'd be different types of undeads, depending on the zone (desertic, swampy, etc...). About evil people, or human enemies, I've thought about it too, but the only thing I was able to come with was the South Pillagers.

Korath could be interesting, it'd be another intelligent species, with the human, elves and dwarves. Maybe it could be a civilization living in the Badlands? Or a new island?

About new islands, I've already came up with an idea: the Shire Island, the only story about it for now is that a 'Green Dragon' called Shire Flying Beast would be sleeping here for decades. He'd be the last of his species (the 'Green Dragons').
